File: assets/lua/meta/vector.lua
--- @meta
--- 
--- @class vector
vector = {}

--- 
--- @param inOutA math.Vector4
--- @param inOutB math.Vector4
--- @param c math.Vector4
--- @overload fun(inOutA: math.Vector, inOutB: math.Vector, c: math.Vector): 
--- @overload fun(inOutA: math.Vector2, inOutB: math.Vector2, c: math.Vector2): 
--- @overload fun(inOutA: math.Vector4, inOutB: math.Vector4): 
--- @overload fun(inOutA: math.Vector, inOutB: math.Vector): 
--- @overload fun(inOutA: math.Vector2, inOutB: math.Vector2): 
--- @overload fun(inOutA: vector.Vector, inOutB: vector.Vector, c: vector.Vector): 
--- @overload fun(inOutA: vector.Vector, inOutB: vector.Vector): 
function vector.to_min_max(inOutA, inOutB, c) end

--- 
--- @param str string
--- @return math.Vector ret0
--- @overload fun(str: string): math.Vector
function vector.create_from_string(str) end

--- 
--- @return math.Vector ret0
--- @overload fun(): math.Vector
function vector.random() end

--- 
--- @param p math.Vector
--- @param d math.Vector
--- @return math.Vector ret0_1
--- @return math.Vector ret0_2
--- @overload fun(p: math.Vector, d: math.Vector, interocularDistance: number): math.Vector, math.Vector
--- @overload fun(p: math.Vector, d: math.Vector, interocularDistance: number, convergenceDistance: number): math.Vector, math.Vector
--- @overload fun(p: vector.Vector, d: vector.Vector): math.Vector, math.Vector
--- @overload fun(p: vector.Vector, d: vector.Vector, interocularDistance: number): math.Vector, math.Vector
--- @overload fun(p: vector.Vector, d: vector.Vector, interocularDistance: number, convergenceDistance: number): math.Vector, math.Vector
function vector.calc_spherical_stereo_transform(p, d) end

--- 
--- @param t 
--- @param outMin math.Vector4
--- @param outMax math.Vector4
--- @overload fun(t: , outMin: math.Vector, outMax: math.Vector): 
--- @overload fun(t: , outMin: math.Vector2, outMax: math.Vector2): 
--- @overload fun(t: , outMin: vector.Vector, outMax: vector.Vector): 
function vector.get_min_max(t, outMin, outMax) end

--- 
--- @return math.Vector2 ret0
function vector.random2D() end

--- 
--- @param points 
--- @return math.Vector ret0
--- @overload fun(points: ): math.Vector
function vector.calc_average(points) end

--- 
--- @param refPos math.Vector
--- @param angVel math.Vector
--- @param tgtPos math.Vector
--- @return math.Vector ret0
--- @overload fun(refPos: vector.Vector, angVel: vector.Vector, tgtPos: vector.Vector): math.Vector
function vector.calc_linear_velocity_from_angular(refPos, angVel, tgtPos) end

--- 
--- @param points 
--- @param ang number
--- @param outNormal math.Vector
--- @param outDistance number
--- @overload fun(points: , ang: number, outNormal: vector.Vector, outDistance: number): 
function vector.calc_best_fitting_plane(points, ang, outNormal, outDistance) end


--- 
--- @class vector.Vector
--- @field g number 
--- @field b number 
--- @field r number 
--- @field z number 
--- @field x number 
--- @field y number 
--- @overload fun(args: string):vector.Vector
--- @overload fun(arg1: math.Vector2, arg2: number):vector.Vector
--- @overload fun(arg1: number, arg2: number, arg3: number):vector.Vector
--- @overload fun():vector.Vector
vector.Vector = {}

--- 
--- @param b vector.Vector
--- @return math.Quaternion ret0
function vector.Vector:GetRotation(b) end

--- 
--- @param b math.ScaledTransform
--- @overload fun(b: math.Transform): 
function vector.Vector:Mul(b) end

--- 
--- @param 1 vector.Vector
function vector.Vector:__eq(arg1) end

--- 
--- @return math.EulerAngles ret0
function vector.Vector:ToEulerAngles() end

--- 
--- @return math.Mat4 ret0
function vector.Vector:ToMatrix() end

--- 
--- @return number ret0
function vector.Vector:Length() end

--- 
--- @param n vector.Vector
--- @param d number
function vector.Vector:Reflect(n, d) end

--- 
--- @param orientation math.Quaternion
--- @overload fun(normal: vector.Vector, angle: number): 
--- @overload fun(ang: math.EulerAngles): 
function vector.Vector:Rotate(orientation) end

--- 
--- @return vector.Vector ret0
function vector.Vector:GetNormal() end

--- 
function vector.Vector:Normalize() end

--- 
--- @return number ret0
function vector.Vector:LengthSqr() end

--- 
--- @param b vector.Vector
--- @return number ret0
function vector.Vector:Distance(b) end

--- 
--- @param p1 vector.Vector
--- @return number ret0
function vector.Vector:DistanceSqr(p1) end

--- 
--- @param p1 vector.Vector
--- @param n vector.Vector
--- @return number ret0
function vector.Vector:PlanarDistance(p1, n) end

--- 
--- @param b vector.Vector
--- @return number ret0
function vector.Vector:GetAngle(b) end

--- 
--- @param n vector.Vector
--- @return vector.Vector ret0
function vector.Vector:Project(n) end

--- 
--- @param p1 vector.Vector
--- @param n vector.Vector
--- @return number ret0
function vector.Vector:PlanarDistanceSqr(p1, n) end

--- 
--- @param 1 vector.Vector
function vector.Vector:__sub(arg1) end

--- 
--- @param 1 vector.Vector
--- @overload fun(arg1: vector.Vector): 
--- @overload fun(arg1: math.Quaternion): 
--- @overload fun(arg1: vector.Vector): 
--- @overload fun(arg1: number): 
function vector.Vector:__mul(arg1) end

--- 
--- @return number ret0
function vector.Vector:GetPitch() end

--- 
--- @param 1 vector.Vector
--- @overload fun(arg1: number): 
function vector.Vector:__div(arg1) end

--- 
--- @param origin vector.Vector
--- @param ang math.EulerAngles
function vector.Vector:RotateAround(origin, ang) end

--- 
--- @param vecB vector.Vector
--- @param factor number
--- @return vector.Vector ret0
function vector.Vector:Lerp(vecB, factor) end

--- 
function vector.Vector:__unm() end

--- 
--- @param vp math.Mat4
--- @return math.Vector2 ret0
function vector.Vector:ToScreenUv(vp) end

--- 
function vector.Vector:__tostring() end

--- 
--- @return number ret0
function vector.Vector:GetYaw() end

--- 
--- @param n vector.Vector
--- @param d number
--- @return vector.Vector ret0
function vector.Vector:ProjectToPlane(n, d) end

--- 
--- @return vector.Vector ret0
function vector.Vector:GetPerpendicular() end

--- 
--- @param v1 vector.Vector
--- @return math.Mat3 ret0
function vector.Vector:OuterProduct(v1) end

--- 
--- @param gridSize int
--- @overload fun(): 
function vector.Vector:SnapToGrid(gridSize) end

--- 
--- @param idx int
function vector.Vector:Get(idx) end

--- 
--- @param b vector.Vector
--- @param factor number
function vector.Vector:Slerp(b, factor) end

--- 
--- @param idx int
--- @param val number
--- @overload fun(x: number, y: number, z: number): 
--- @overload fun(vecB: vector.Vector): 
function vector.Vector:Set(idx, val) end

--- 
--- @return vector.Vector ret0
function vector.Vector:Copy() end

--- 
--- @param 1 vector.Vector
function vector.Vector:__add(arg1) end

--- 
--- @param b vector.Vector
--- @return number ret0
function vector.Vector:DotProduct(b) end

--- 
--- @param vecB vector.Vector
--- @return vector.Vector ret0
function vector.Vector:Cross(vecB) end

--- 
--- @param b vector.Vector
--- @overload fun(b: vector.Vector, epsilon: number): 
function vector.Vector:Equals(b) end

File: core/shared/src/implementation/lua/classes/vector.cpp
// SPDX-FileCopyrightText: (c) 2019 Silverlan <opensource@pragma-engine.com>
// SPDX-License-Identifier: MIT
module;

module pragma.shared;

import :scripting.lua.classes.vector;

pragma::math::Vertex Lua::Vertex::Copy(pragma::math::Vertex &v) { return v; }
pragma::math::VertexWeight Lua::VertexWeight::Copy(pragma::math::VertexWeight &vw) { return vw; }
Vector3i Lua::Vectori::Copy(Vector3i &v) { return v; }
Vector2i Lua::Vector2i::Copy(::Vector2i &v) { return v; }
Vector4i Lua::Vector4i::Copy(::Vector4i &v) { return v; }

void Lua::Vector::Normalize(Vector3 &vec) { uvec::normalize(&vec); }

static const Vector3 vPitch(1, 0, 0);
static const Vector3 vYaw(0, 1, 0);
static const Vector3 vRoll(0, 0, 1);
void Lua::Vector::Rotate(lua::State *, Vector3 &vec, const EulerAngles &ang) { uvec::rotate(&vec, ang); }
void Lua::Vector::Rotate(lua::State *, Vector3 &vec, const Vector3 &normal, float angle) { vec = glm::gtc::rotate(vec, angle, normal); }
void Lua::Vector::Rotate(lua::State *, Vector3 &vec, const Quat &orientation) { uvec::rotate(&vec, orientation); }
void Lua::Vector::RotateAround(lua::State *, Vector3 &vec, const Vector3 &origin, const EulerAngles &ang) { uvec::rotate_around(&vec, ang, origin); }
Vector3 Lua::Vector::Copy(lua::State *l, const Vector3 &vec) { return vec; }
Vector3 Lua::Vector::Lerp(lua::State *l, const Vector3 &vec, const Vector3 &vecB, float factor) { return uvec::lerp(vec, vecB, factor); }

void Lua::Vector::Set(lua::State *, Vector3 &vec, const Vector3 &vecB)
{
	vec.x = vecB.x;
	vec.y = vecB.y;
	vec.z = vecB.z;
}

void Lua::Vector::Set(lua::State *, Vector3 &vec, float x, float y, float z)
{
	vec.x = x;
	vec.y = y;
	vec.z = z;
}

Mat4 Lua::Vector::ToMatrix(lua::State *l, const Vector3 &vec) { return glm::gtc::translate(umat::identity(), vec); }

void Lua::Vector::SnapToGrid(lua::State *, Vector3 &vec, UInt32 gridSize) { uvec::snap_to_grid(vec, gridSize); }
void Lua::Vector::SnapToGrid(lua::State *l, Vector3 &vec) { SnapToGrid(l, vec, 1); }

////////////////////////

Vector2 Lua::Vector2::GetNormal(const ::Vector2 &vec)
{
	Vector3 n(vec.x, vec.y, 0);
	uvec::normalize(&n);
	return ::Vector2(n.x, n.y);
}
void Lua::Vector2::Normalize(::Vector2 &vec)
{
	Vector3 n(vec.x, vec.y, 0);
	uvec::normalize(&n);
	vec.x = n.x;
	vec.y = n.y;
}
float Lua::Vector2::Length(const ::Vector2 &vec) { return uvec::length(Vector3(vec.x, vec.y, 0)); }
float Lua::Vector2::LengthSqr(const ::Vector2 &vec) { return uvec::length_sqr(Vector3(vec.x, vec.y, 0)); }
float Lua::Vector2::Distance(const ::Vector2 &a, const ::Vector2 &b)
{
	Vector3 v = Vector3(b.x, b.y, 0) - Vector3(a.x, a.y, 0);
	return uvec::length(v);
}
float Lua::Vector2::DistanceSqr(const ::Vector2 &a, const ::Vector2 &b)
{
	Vector3 v = Vector3(b.x, b.y, 0) - Vector3(a.x, a.y, 0);
	return uvec::length_sqr(v);
}
Vector2 Lua::Vector2::Cross(lua::State *l, const ::Vector2 &a, const ::Vector2 &b)
{
	Vector3 cross = glm::cross(Vector3(a.x, a.y, 0), Vector3(b.x, b.y, 0));
	return ::Vector2(cross.x, cross.y);
}
float Lua::Vector2::DotProduct(lua::State *l, const ::Vector2 &a, const ::Vector2 &b) { return glm::dot(Vector3(a.x, a.y, 0), Vector3(b.x, b.y, 0)); }
void Lua::Vector2::Rotate(lua::State *, ::Vector2 &vec, const EulerAngles &ang)
{
	Vector3 v = Vector3(vec.x, vec.y, 0);
	uvec::rotate(&v, ang);
	vec.x = v.x;
	vec.y = v.y;
}
void Lua::Vector2::RotateAround(lua::State *, ::Vector2 &vec, const ::Vector2 &origin, const EulerAngles &ang)
{
	Vector3 v(vec.x, vec.y, 0);
	Vector3 vB = Vector3(origin.x, origin.y, 0);
	uvec::rotate_around(&v, ang, vB);
	vec.x = v.x;
	vec.y = v.y;
}
Vector2 Lua::Vector2::Copy(lua::State *l, const ::Vector2 &vec) { return ::Vector2(vec.x, vec.y); }
Vector2 Lua::Vector2::Lerp(lua::State *l, const ::Vector2 &vec, const ::Vector2 &vecB, float factor)
{
	Vector3 v = uvec::lerp(Vector3(vec.x, vec.y, 0), Vector3(vecB.x, vecB.y, 0), factor);
	return ::Vector2 {v.x, v.y};
}
void Lua::Vector2::Set(lua::State *, ::Vector2 &vec, const ::Vector2 &vecB)
{
	vec.x = vecB.x;
	vec.y = vecB.y;
}
void Lua::Vector2::Set(lua::State *, ::Vector2 &vec, float x, float y)
{
	vec.x = x;
	vec.y = y;
}
Vector2 Lua::Vector2::Project(lua::State *l, const ::Vector2 &vec, const ::Vector2 &n) { return glm::gtx::proj(vec, n); }

////////////////////////

Vector4 Lua::Vector4::GetNormal(const ::Vector4 &vec) { return glm::normalize(vec); }
void Lua::Vector4::Normalize(::Vector4 &vec) { vec = glm::normalize(vec); }
float Lua::Vector4::Length(const ::Vector4 &vec) { return glm::length(vec); }
float Lua::Vector4::LengthSqr(const ::Vector4 &vec) { return glm::gtx::length2(vec); }
float Lua::Vector4::Distance(const ::Vector4 &a, const ::Vector4 &b) { return glm::distance(a, b); }
float Lua::Vector4::DistanceSqr(const ::Vector4 &a, const ::Vector4 &b) { return glm::gtx::distance2(a, b); }
float Lua::Vector4::DotProduct(lua::State *l, const ::Vector4 &a, const ::Vector4 &b) { return glm::dot(a, b); }
Vector4 Lua::Vector4::Copy(lua::State *l, const ::Vector4 &vec) { return ::Vector4(vec.x, vec.y, vec.z, vec.w); }
Vector4 Lua::Vector4::Lerp(lua::State *l, const ::Vector4 &vec, const ::Vector4 &vecB, float factor) { return vec + (vecB - vec) * factor; }
void Lua::Vector4::Set(lua::State *, ::Vector4 &vec, const ::Vector4 &vecB)
{
	vec.w = vecB.w;
	vec.x = vecB.x;
	vec.y = vecB.y;
	vec.z = vecB.z;
}
void Lua::Vector4::Set(lua::State *, ::Vector4 &vec, float x, float y, float z, float w)
{
	vec.w = w;
	vec.x = x;
	vec.y = y;
	vec.z = z;
}
Vector4 Lua::Vector4::Project(lua::State *l, const ::Vector4 &vec, const ::Vector4 &n) { return glm::gtx::proj(vec, n); }

////////////////////////

template<class TVector>
static void to_min_max(TVector &inOutA, TVector &inOutB)
{
	auto numComponents = inOutA.length();
	for(auto i = decltype(numComponents) {0}; i < numComponents; ++i) {
		if(inOutB[i] < inOutA[i]) {
			auto t = inOutA[i];
			inOutA[i] = inOutB[i];
			inOutB[i] = t;
		}
	}
}

template<class TVector>
static void to_min_max(TVector &inOutA, TVector &inOutB, const TVector &c)
{
	auto numComponents = inOutA.length();
	for(auto i = decltype(numComponents) {0}; i < numComponents; ++i) {
		if(c[i] < inOutA[i])
			inOutA[i] = c[i];
		if(c[i] > inOutB[i])
			inOutB[i] = c[i];
	}
}

void Lua::vector::to_min_max(::Vector2 &inOutA, ::Vector2 &inOutB) { ::to_min_max(inOutA, inOutB); }
void Lua::vector::to_min_max(Vector3 &inOutA, Vector3 &inOutB) { ::to_min_max(inOutA, inOutB); }
void Lua::vector::to_min_max(::Vector4 &inOutA, ::Vector4 &inOutB) { ::to_min_max(inOutA, inOutB); }
void Lua::vector::to_min_max(::Vector2 &inOutA, ::Vector2 &inOutB, const ::Vector2 &c) { ::to_min_max(inOutA, inOutB, c); }
void Lua::vector::to_min_max(Vector3 &inOutA, Vector3 &inOutB, const Vector3 &c) { ::to_min_max(inOutA, inOutB, c); }
void Lua::vector::to_min_max(::Vector4 &inOutA, ::Vector4 &inOutB, const ::Vector4 &c) { ::to_min_max(inOutA, inOutB, c); }

template<class TVector>
static void get_min_max(lua::State *l, luabind::table<> t, TVector &outMin, TVector &outMax)
{
	auto numEls = Lua::GetObjectLength(l, 1);
	if(numEls == 0) {
		outMin = {};
		outMax = {};
		return;
	}
	for(auto i = decltype(outMin.length()) {0u}; i < outMin.length(); ++i) {
		outMin[i] = std::numeric_limits<float>::max();
		outMax[i] = std::numeric_limits<float>::lowest();
	}
	for(auto it = luabind::iterator {t}; it != luabind::iterator {}; ++it) {
		auto val = luabind::object_cast_nothrow<TVector>(*it, TVector {});
		for(auto i = decltype(outMin.length()) {0u}; i < outMin.length(); ++i) {
			outMin[i] = pragma::math::min(outMin[i], val[i]);
			outMax[i] = pragma::math::min(outMax[i], val[i]);
		}
	}
}

void Lua::vector::get_min_max(lua::State *l, luabind::table<> t, ::Vector2 &outMin, ::Vector2 &outMax) { return ::get_min_max<::Vector2>(l, t, outMin, outMax); }
void Lua::vector::get_min_max(lua::State *l, luabind::table<> t, Vector3 &outMin, Vector3 &outMax) { return ::get_min_max<Vector3>(l, t, outMin, outMax); }
void Lua::vector::get_min_max(lua::State *l, luabind::table<> t, ::Vector4 &outMin, ::Vector4 &outMax) { return ::get_min_max<::Vector4>(l, t, outMin, outMax); }

Vector2 Lua::vector::random_2d()
{
	float azimuth = pragma::math::random(0.f, 2.f) * static_cast<float>(pragma::math::pi);
	return ::Vector2(std::cos(azimuth), std::sin(azimuth));
}

Vector3 Lua::vector::calc_average(luabind::table<> points)
{
	Vector3 avg {};
	uint32_t numPoints = 0;
	for(auto it = luabind::iterator {points}; it != luabind::iterator {}; ++it) {
		auto val = luabind::object_cast_nothrow<Vector3>(*it, Vector3 {});
		avg += val;
		++numPoints;
	}
	avg /= static_cast<float>(numPoints);
	return avg;
}

void Lua::vector::calc_best_fitting_plane(luabind::table<> points, float ang, Vector3 &outNormal, double &outDistance)
{
	auto avg = calc_average(points);
	auto mat = umat::create_from_axis_angle(avg, ang);

	pragma::math::calc_best_fitting_plane(mat, avg, outNormal, outDistance);
}
